src/realm/object-store/audit.hpp

```diff
@@ -16,20 +16,109 @@
+    // Error handler which is called if fatal sync errors occur on the sync
+    // Realm. If null, an error is logged then abort() is called.
+    std::function<void(SyncError)> sync_error_handler;
+    // Metadata to attach to each audit event. Each key used must be a property
+    // in the server-side schema for AuditEvent. This is not validated and will
+    // result in a sync error if violated.
+    std::vector<std::pair<std::string, std::string>> metadata;
+};
 
 class AuditInterface {
 public:
-    virtual ~AuditInterface() {}
+    virtual ~AuditInterface() = default;
+
+    // Internal interface for recording auditable events. SDKs may need to call
+    // record_read() if they do not go through Object; record_query() and
+    // record_write() should be handled automatically
+    virtual void record_query(VersionID, const TableView&) = 0;
+    virtual void record_read(VersionID, const Obj& obj, const Obj& parent, ColKey col) = 0;
+    virtual void record_write(VersionID old_version, VersionID new_version) = 0;
+
+    // -- Audit functionality which should be exposed in the SDK
+
+    // Update the metadata attached to subsequence audit events. Does not effect
+    // the current audit scope if called while a scope is active.
+    virtual void update_metadata(std::vector<std::pair<std::string, std::string>> new_metadata) = 0;
 
-    virtual void record_query(realm::VersionID, realm::TableView const&) = 0;
-    virtual void record_read(realm::VersionID, realm::RowExpr) = 0;
-    virtual void record_write(realm::VersionID, realm::VersionID) = 0;
+    // Begin an audit scope. The given `name` is stored in the activity field
+    // of each generated event.
+    virtual void begin_scope(std::string_view name) = 0;
+    // End the current scope and asynchronously save it to disk. The optional
+    // completion function is called once it has been committed (or an error
+    // ocurred while trying to do so).
+    virtual void end_scope(util::UniqueFunction<void(std::exception_ptr)>&& completion = nullptr) = 0;
+    // Record a custom audit event. Does not use the scope (and does not need to be inside a scope).
+    virtual void record_event(std::string_view activity, util::Optional<std::string> event_type,
+                              util::Optional<std::string> data,
+                              util::UniqueFunction<void(std::exception_ptr)>&& completion) = 0;
+
+    // -- Test helper functionality
+
+    // Wait for all scopes to be written to disk. Does not wait for them to be
+    // uploaded to the server.
+    virtual void wait_for_completion() = 0;
+    // Wait for there to be no more data to upload. This is not a precise check;
+    // if more scopes are created while this is waiting they may or may not be
+    // included in the wait.
+    virtual void wait_for_uploads() = 0;
 };
+
+std::shared_ptr<AuditInterface> make_audit_context(std::shared_ptr<DB>, RealmConfig const& parent_config);
```

Comment:
@tgoyne  we are missing the definition of this method when the platform is APPLE, this won't compile https://ci.realm.io/blue/organizations/jenkins/realm%2Frealm-core/detail/PR-5495/7/pipeline/